bin/common.c

```diff
@@ -75,3 +77,60 @@ int key_log_callback(void *file, struct s2n_connection *conn, uint8_t *logline,
 
     return fflush((FILE *)file);
 }
+
+static int s2n_get_psk_hmac_alg(s2n_psk_hmac *psk_hmac, char *hmac_str)
+{
+    GUARD_EXIT_NULL(psk_hmac, "PSK HMAC is NULL");
+    GUARD_EXIT_NULL(hmac_str, "HMAC string is NULL");
+
+    if (strcmp(hmac_str, "S2N_PSK_HMAC_SHA256") == 0) {
+        *psk_hmac = S2N_PSK_HMAC_SHA256;
+    } else if (strcmp(hmac_str, "S2N_PSK_HMAC_SHA384") == 0) {
+        *psk_hmac = S2N_PSK_HMAC_SHA384;
+    } else {
+        return S2N_FAILURE;
+    }
+    return S2N_SUCCESS;
+}
+
+int s2n_setup_external_psk(struct s2n_psk *psk_list[S2N_MAX_PSK_LIST_LENGTH], size_t *psk_idx, char *params)
+{
+    GUARD_EXIT_NULL(psk_list, "PSK list is NULL");
+    GUARD_EXIT_NULL(psk_idx, "PSK index is NULL");
+    GUARD_EXIT_NULL(params, "PSK parameters is NULL");
+
+    struct s2n_psk *psk = s2n_external_psk_new();
+    GUARD_EXIT_NULL(psk, "Error initializing PSK, NULL value obtained");
+    /* Default HMAC algorithm is S2N_PSK_HMAC_SHA256 */
+    s2n_psk_hmac psk_hmac_alg = S2N_PSK_HMAC_SHA256;
+    size_t idx = 0;
+    for (char *token = strtok(params, ","); token != NULL; token = strtok(NULL, ","), idx++) {
+        unsigned char *secret = NULL;
+        long secret_len = 0;
+        switch (idx) {
+            case 0:
+                GUARD_EXIT(s2n_psk_set_identity(psk, (const uint8_t *)token, strlen(token)),
+                             "Error setting psk identity");
+                break;
+            case 1:
+                /* Ref: https://www.openssl.org/docs/man1.1.0/man3/OPENSSL_hexstr2buf.html
+                 * OPENSSL_hexstr2buf() parses str as a hex string and returns a pointer to the parsed value. 
+                 * The memory is allocated by calling OPENSSL_malloc() and should be released by calling OPENSSL_free().
+                 */
+                secret = OPENSSL_hexstr2buf((const char *)token, &secret_len);
+                GUARD_EXIT_NULL(secret, "Error converting hex-encoded secret value to bytes");
+                GUARD_EXIT(s2n_psk_set_secret(psk, (const uint8_t *)secret, secret_len), "Error setting psk secret");
```

Comment:
If this is so difficult and requires an OPENSSL method, maybe S2N needs a `s2n_psk_set_secret_hex` method?